### PR TITLE
move from deployment to statefulset

### LINKED ISSUE #364: switch the provider components from deployment to statefulset
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

https://github.com/kubernetes-sigs/cluster-api/issues/983
for more info
**What did you expect to happen:**

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api version:
- Minikube/KIND version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by jichenjc**
per discussion , deployment has sideeffect such as potential
race condition, so need switch to statefulset.

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #364 

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by jichenjc** [comment]
this will be updated after kustomize patch complete as the test rely on that PR

**Comment by jichenjc** [comment]
the statefulsets created successfully
```
root@jichen1:/home/cloudusr/go/src/sigs.k8s.io/cluster-api-provider-openstack/cmd/clusterctl/examples/openstack# kubectl get pods --all-namespaces
NAMESPACE                   NAME                                               READY   STATUS    RESTARTS   AGE
kube-system                 coredns-86c58d9df4-5slz8                           1/1     Running   0          7m55s
kube-system                 coredns-86c58d9df4-chctx                           1/1     Running   0          7m55s
kube-system                 etcd-clusterapi-control-plane                      1/1     Running   0          7m9s
kube-system                 kube-apiserver-clusterapi-control-plane            1/1     Running   0          6m59s
kube-system                 kube-controller-manager-clusterapi-control-plane   1/1     Running   0          7m3s
kube-system                 kube-proxy-6qrgc                                   1/1     Running   0          7m55s
kube-system                 kube-scheduler-clusterapi-control-plane            1/1     Running   0          7m5s
kube-system                 weave-net-k7vfz                                    2/2     Running   1          7m55s
openstack-provider-system   clusterapi-controllers-0                           1/1     Running   0          7m54s
system                      controller-manager-0                               1/1     Running   0          7m54s
root@jichen1:/home/cloudusr/go/src/sigs.k8s.io/cluster-api-provider-openstack/cmd/clusterctl/examples/openstack# kubectl get statefulsets --all-namespaces
NAMESPACE                   NAME                     READY   AGE
openstack-provider-system   clusterapi-controllers   1/1     8m10s
system                      controller-manager       1/1     8m10s
root@jichen1:/home/cloudusr/go/src/sigs.k8s.io/cluster-api-provider-openstack/cmd/clusterctl/examples/openstack# kubectl get deployments --all-namespaces
NAMESPACE     NAME      READY   UP-TO-DATE   AVAILABLE   AGE
kube-system   coredns   2/2     2            2           8m25s
```

File: config/manager/statefulset.yaml [hunk 1]
```diff
@@ -12,6 +12,7 @@ spec:
     matchLabels:
       control-plane: controller-manager
       controller-tools.k8s.io: "1.0"
+  serviceName: controller-manager-service
```
Comment (hidekazuna) [inline config/manager/statefulset.yaml]: Service name is openstack-provider-controller-manager-service.
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/master/config/manager/service.yaml#L7

**Comment by hidekazuna** [comment]
@jichenjc Thanks for fixing.

**Comment by jichenjc** [comment]
@chrigl @hidekazuna can you help to take a look at this ? this currently blocked the creation.....

**Pull request merged**